addons/crm/models/crm_lead.py

```diff
@@ -860,40 +911,52 @@ def search_fetch(self, domain, field_names, offset=0, limit=None, order=None):
-        - From won to lost : We need to decrement corresponding won count + increment corresponding lost count
-        in scoring frequency table."""
+    def _handle_won_lost(self, old_status_by_lead, new_status_by_lead):
+        """ This method handles all changes of won / lost status of leads on creation / writing,
+        and update the scoring frequency table accordingly:
+        - To lost : Increment corresponding lost count
+        - To won : Increment corresponding won count
+        - Leaving lost : Decrement corresponding lost count
+        - Leaving won : Decrement corresponding won count
+        More than one operation can happen simultaneously, for instance, going from lost to won:
+        Decrement corresponding lost count + increment corresponding won count.
+
+        A lead is WON when in won stage (and probability = 100% but that is implied and constrained)
+        A lead is LOST when active = False AND probability = 0
+        In every other case, the lead is not won nor lost.
+
+        :param old_status_by_lead: dict of old status by lead: {lead.id: {'is_lost': ..., 'is_won': ...}}
+        :param new_status_by_lead: dict of new status by lead: {lead.id: {'is_lost': ..., 'is_won': ...}}
+        """
         Lead = self.env['crm.lead']
-        leads_reach_won = Lead
-        leads_leave_won = Lead
-        leads_reach_lost = Lead
-        leads_leave_lost = Lead
-        won_stage_ids = self.env['crm.stage'].search([('is_won', '=', True)]).ids
-        for lead in self:
-            if 'stage_id' in vals:
-                if vals['stage_id'] in won_stage_ids:
-                    if lead.probability == 0:
-                        leads_leave_lost += lead
-                    leads_reach_won += lead
-                elif lead.stage_id.id in won_stage_ids and lead.active:  # a lead can be lost at won_stage
-                    leads_leave_won += lead
-            if 'active' in vals:
-                if not vals['active'] and lead.active:  # archive lead
-                    if lead.stage_id.id in won_stage_ids and lead not in leads_leave_won:
-                        leads_leave_won += lead
-                    leads_reach_lost += lead
-                elif vals['active'] and not lead.active:  # restore lead
-                    leads_leave_lost += lead
-
-        leads_reach_won._pls_increment_frequencies(to_state='won')
-        leads_leave_won._pls_increment_frequencies(from_state='won')
-        leads_reach_lost._pls_increment_frequencies(to_state='lost')
-        leads_leave_lost._pls_increment_frequencies(from_state='lost')
+        leads_reach_won_ids = []
+        leads_leave_won_ids = []
+        leads_reach_lost_ids = []
+        leads_leave_lost_ids = []
+
+        for lead_id, new_status in new_status_by_lead.items():
+            old_status = old_status_by_lead.get(lead_id, {'is_lost': False,
+                                                          'is_won': False})
```

Comment:
Would use "self" to be a real recordset based method, like
```suggestion
        for lead in self:
            new_status = new_status_by_lead.get(
                lead.id, {'is_lost': False, 'is_won': False}
            )
            old_status = old_status_by_lead.get(
                lead.id, {'is_lost': False, 'is_won': False}
            )
```
```